src/sp/protocol/mqtt/nmq_mqtt.c

```diff
@@ -831,21 +827,35 @@ nano_pipe_close(void *arg)
 		clientid = (char *) conn_param_get_clientid(p->conn_param);
 	}
 	if (clientid) {
-		clientid_key = DJBHashn(clientid, strlen(clientid));
-		nni_id_set(&s->cached_sessions, clientid_key, p);
-		nni_mtx_lock(&p->lk);
-		// set event to false avoid of sending the disconnecting msg
-		p->event                   = false;
-		npipe->cache               = true;
-		p->conn_param->clean_start = 1;
-		nni_atomic_swap_bool(&npipe->p_closed, false);
-		if (nni_list_active(&s->recvpipes, p)) {
-			nni_list_remove(&s->recvpipes, p);
+		nni_pipe *new_pipe;
+		nni_pipe_find(&new_pipe, npipe->p_id);
+		if (new_pipe == npipe) {
+			nni_id_set(&s->cached_sessions, npipe->p_id, p);
+			nni_mtx_lock(&p->lk);
+			// set event to false avoid of sending the
+			// disconnecting msg
+			p->event                   = false;
+			npipe->cache               = true;
+			p->conn_param->clean_start = 1;
+			nni_atomic_swap_bool(&npipe->p_closed, false);
+			if (nni_list_active(&s->recvpipes, p)) {
+				nni_list_remove(&s->recvpipes, p);
+			}
+			nano_nni_lmq_flush(&p->rlmq, false);
+			nni_mtx_unlock(&s->lk);
+			nni_mtx_unlock(&p->lk);
+			return -1;
 		}
-		nano_nni_lmq_flush(&p->rlmq, false);
-		nni_mtx_unlock(&s->lk);
-		nni_mtx_unlock(&p->lk);
-		return -1;
+		// take params from npipe to new pipe
+		new_pipe->packet_id = npipe->packet_id;
+		// there should be no msg in this map
+		new_pipe->nano_qos_db = npipe->nano_qos_db;
+		npipe->nano_qos_db = NULL;
+
+		nni_list *l        = new_pipe->subinfol;
+		new_pipe->subinfol = npipe->subinfol;
+		npipe->subinfol    = l;
+		log_info("client kick itself while keeping session!");
```

Comment:
But nni_pipe_close will not be called when p->cache==true.
So what's the situation?